master/pkg/stream/stream.go

```diff
@@ -0,0 +1,205 @@
+package stream
+
+import (
+	"encoding/json"
+	"sync"
+)
+
+// Msg is an object with a message and a sequence number and json marshal caching.
+type Msg interface {
+	SeqNum() int64
+	UpsertMsg() UpsertMsg
+	DeleteMsg() DeleteMsg
+}
+
+// Event contains the old and new version a Msg.  Inserts will have Before==nil, deletions will
+// have After==nil.
+type Event[T Msg] struct {
+	Before *T `json:"before"`
+	After  *T `json:"after"`
+
+	upsertCache interface{}
+	deleteCache interface{}
+}
+
+// PreparableMessage is an intermediary message that is ready to be marshaled and broadcast.
+type PreparableMessage interface{}
```

Comment:
non-blocking: I remember writing this and I think I may have misused interfaces here... This should work and doesn't impact the functionality, but I think this should be changed in the future. As it stands, anything can be a `PreparableMessage`, not just upsert and delete msgs

Reply:
All good I agree that this should probably be more restrictive in terms of definition

Reply:
Made this more restrictive by requiring the `MarshalJSON()` to be implemented. I also changed the name of this interface to `MarshallableMessage` so that it's not as easily confused with `websocket.PreparedMessage`